app/Models/User.php

```diff
@@ -589,6 +589,65 @@ public function scopeHasToPayKKTNetregInSemester(Builder $query, int $semester_i
             });
     }
 
+    /**
+     * Scope a query to only include users who got accepted in the specified year.
+     */
+    public function scopeYearOfAcceptance(Builder $query, int $yearOfAcceptance): Builder
+    {
+        return $query->whereHas('educationalInformation', function (Builder $query) use ($yearOfAcceptance) {
+            $query->where('year_of_acceptance', $yearOfAcceptance);
+        });
+    }
+
+    /**
+     * Scope a query to only include users whose name contains the given string.
+     */
+    public function scopeNameLike(Builder $query, string $nameLike): Builder
+    {
+        return $query->where('name', 'like', '%' . $nameLike . '%');
+    }
+
+    /**
+     * Scope a query to only include users who have all the specified roles.
+     * The roles are specified by their IDs.
+     */
+    public function scopeHasAllRoleIds(Builder $query, array $roleIdsAll): void
+    {
+        foreach ($roleIdsAll as $roleId) {
+            $query->whereHas('roles', function (Builder $query) use ($roleId) {
+                $query->where('id', $roleId);
+            });
+        }
+    }
+
+    /**
+     * Scope a query to only include users who are in all the specified workshops.
+     * The workshops are specified by their IDs.
+     */
+    public function scopeInAllWorkshopIds(Builder $query, array $workshopsIdsAll): void
+    {
+        foreach ($workshopsIdsAll as $workshopId) {
+            $query->whereHas('workshops', function (Builder $query) use ($workshopId) {
+                $query->where('id', $workshopId);
```

Comment:
use whereIn and you can pass the $workshopIdsAll without the foreach

Reply:
Done!